rp235x-hal/src/xh3irq.rs

```diff
@@ -0,0 +1,239 @@
+    unsafe {
+        core::arch::asm!(
+            "csrc 0xbe0, {0}",
+            in(reg) mask_index
+        );
+    }
+}
+
+/// Check if an interrupt is enabled
+#[cfg(all(target_arch = "riscv32", target_os = "none"))]
+pub fn is_enabled(irq: rp235x_pac::Interrupt) -> bool {
+    let (index, bits) = interrupt_to_mask(irq);
+    let index = index as u32;
+    let mut csr_rdata: u32;
+    // Do a CSR Read-Set on RVCSR_MEIEA_OFFSET
+    unsafe {
+        core::arch::asm!(
+            "csrrs {0}, 0xbe0, {1}",
+            out(reg) csr_rdata,
+            in(reg) index
+        );
+    }
+    let bitmask = (bits as u32) << 16;
+    (csr_rdata & bitmask) != 0
+}
+
+/// Set an interrupt as pending, even if it isn't.
+#[cfg(all(target_arch = "riscv32", target_os = "none"))]
+pub fn pend(irq: rp235x_pac::Interrupt) {
+    let mask_index = interrupt_to_mask_index(irq);
+    // Do a RISC-V CSR Set on RVCSR_MEIFA_OFFSET
+    unsafe {
+        core::arch::asm!(
+            "csrs 0xbe2, {0}",
+            in(reg) mask_index
+        );
+    }
+}
+
+/// Check which interrupt should be handled next
+///
+/// Also updates the state so next time you call this you'll get a different
+/// answer.
+#[cfg(all(target_arch = "riscv32", target_os = "none"))]
+pub fn get_next_interrupt() -> Option<rp235x_pac::Interrupt> {
+    const NOIRQ: u32 = 0x8000_0000;
+
+    let mut csr_rdata: u32;
+    // Do a CSR Read-Set-Immediate on MEINEXT to set the UPDATE bit
+    unsafe {
+        core::arch::asm!(
+            "csrrsi {0}, 0xbe4, 0x01",
+            out(reg) csr_rdata,
+        );
+    }
+
+    if (csr_rdata & NOIRQ) != 0 {
+        return None;
+    }
+
+    let irq_no = (csr_rdata >> 2) as u8;
+    let irq = rp235x_pac::Interrupt::try_from(irq_no).unwrap();
+    Some(irq)
+}
+
+/// Convert an IRQ into a window selection value and separate a bitmask for that
+/// window.
+pub const fn interrupt_to_mask(irq: rp235x_pac::Interrupt) -> (u16, u16) {
+    let irq = irq as u16;
+    // Select a bank of 16 interrupts out of the 512 total
+    let index = irq / 16;
+    // Which interrupt out of the 16 we've selected
+    let bitmask = 1 << (irq % 16);
+    (index, bitmask)
+}
+
+/// Convert an IRQ into a 32-bit value that selects a window and a bit within
+/// that window.
+pub const fn interrupt_to_mask_index(irq: rp235x_pac::Interrupt) -> u32 {
+    let (index, bits) = interrupt_to_mask(irq);
+    (bits as u32) << 16 | index as u32
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_interrupt_to_mask() {
+        let (index, bitmask) = interrupt_to_mask(rp235x_pac::Interrupt::UART1_IRQ);
+        // 34 -> window 2, bit 3
+        assert_eq!(index, 2);
+        assert_eq!(bitmask, 0b0000_0000_0000_0100);
+    }
+
+    #[test]
+    fn test_interrupt_to_mask_index() {
+        let mask = interrupt_to_mask_index(rp235x_pac::Interrupt::UART1_IRQ);
+        // 34 -> bit 3 | window 2
+        assert_eq!(mask, 0x0004_0002);
+    }
```

Comment:
what does the 34 comment mean here?

Reply:
Oh, UART1_IRQ has the value 34

Reply:
Cleaned up the test and used another IRQ where the bit number and the index have different numeric values